rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc

```diff
@@ -0,0 +1,275 @@
+    try {
+        parse(#Name, escapeName(name));
+    } catch ParseError(_): {
+        throw InvalidName(name);
+    }
+}
+
+void checkCausesDoubleDeclaration(set[Define] currentDefs, set[Define] newDefs) {
+    // Is newName already resolvable from a scope where <current-name> is currently declared?
+    throwIfNotEmpty(CausesDoubleDeclaration, {<nD.defined, cD> | Define nD <- newDefs
+                                                               , loc cD <- currentDefs.defined
+                                                               , isContainedIn(cD, nD.scope)
+                                             });
+}
+
+bool isImplicitDefinition(start[Module] _, rel[loc use, loc def] useDef, Define _: <_, _, _, variableId(), defined, _>) = defined in useDef.use; // use of name at implicit declaration loc
+
+bool isImplicitDefinition(start[Module] m, rel[loc, loc] _, Define _: <_, _, _, patternVariableId(), defined, _>) {
+    visit (m) {
+        case qualifiedName(qn):
+            if (locationOfName(qn) == just(defined)) return true;
+        case multiVariable(qn):
+            if (locationOfName(qn) == just(defined)) return true;
+        case variableBecomes(n, _):
+            if (locationOfName(n) == just(defined)) return true;
+    }
+
+    return false;
+}
+
+default bool isImplicitDefinition(start[Module] _, rel[loc, loc] _, Define _) = false;
+
+void checkCausesCapture(TModel tm, start[Module] m, set[Define] currentDefs, set[loc] currentUses, set[Define] newDefs) {
+    set[Define] newNameImplicitDefs = {def | Define def <- newDefs, isImplicitDefinition(m, tm.useDef, def)};
+    set[Define] newNameExplicitDefs = {def | Define def <- newDefs, !isImplicitDefinition(m, tm.useDef, def)};
+
+    // Will this rename turn an implicit declaration of `newName` into a use of a current declaration?
+    rel[loc use, loc oldDef, loc newDef] implicitDeclBecomesUseOfCurrentDecl =
+        {<nD.defined, nD.defined, cD> | Define nD <- newNameImplicitDefs
+                                      , <loc cS, loc cD> <- currentDefs<scope, defined>
+                                      , isContainedIn(nD.defined, cS)
+        };
+
+    // Will this rename turn a use of <oldName> into a use of an existing declaration of <newName> (shadowing)?
+    rel[loc use, loc oldDef, loc newDef] explicitDeclShadowsCurrentDecl =
+        {<cU, cD.defined, nD.defined> | Define nD <- newNameExplicitDefs
+                                      , loc cU <- currentUses
+                                      , Define cD <- getDefines(tm, cU)
+                                      , isContainedIn(cU, nD.scope)
+                                      , isStrictlyContainedIn(nD.scope, cD.scope)
+        };
+
+    throwIfNotEmpty(CausesCapture, implicitDeclBecomesUseOfCurrentDecl + explicitDeclShadowsCurrentDecl);
+}
+
+void checkEditableDefinitions(set[Define] currentDefs) {
+    bool isEditableLocation(loc l) = l.scheme in {"file", "cwd", "home", "memory", "project"};
```

Comment:
It's not possible to make an accurate list, and also, if a file is readonly on disk, we still cannot write to it. Let's let VS Code worry about updating files or not, as it's a whole mess in the virtual file system if something is readonly or not.

Reply:
Yeah, I like the thought, but in the end, I don't think we can get this water tight, without adding a `isWritable(loc)` function to `IO.rsc` (not even sure if we can do that properly)

Reply:
Different approach (not for this PR, but for cross-module features that we're working on next): reject the renaming if any related definition is outside of the current workspace?